File: leetcode/strings/valid-palindrome-two-pointers-from-outside.js
// 2 pointers from the outside moving towards the center
const isValidPalindromeOutside = function(s) {
  s = s.replace(/[^A-Za-z0-9]/g, "").toLowerCase();

  // Initialize left/right pointers at start and end of string
  // respectively.
  let left = 0, right = s.length - 1;

  // Loop through string characters while comparing them, then
  // move the pointers closer to the center.
  // If the string is odd , the left/right pointers will end
  // at the index immediate to the left and right of the center.
  // There's no need to compare the center character.
  // If the string is even, the left/right pointers will end
  // at the 2 center characters respectively, and we'll need to
  // compare them.
  while (left < right) {
    if (s[left] !== s[right]) {
      return false;
    }
    left++;
    right--;
  }

  return true;
}

try {
  Error.stackTraceLimit = 1;
  if (!(isValidPalindromeOutside("aabaa") === true)) throw new Error("Test Result Error");
  if (!(isValidPalindromeOutside("aabbaa") === true)) throw new Error("Test Result Error");
  if (!(isValidPalindromeOutside("abc") === false)) throw new Error("Test Result Error");
  if (!(isValidPalindromeOutside("a") === true)) throw new Error("Test Result Error");
  if (!(isValidPalindromeOutside("") === true)) throw new Error("Test Result Error");
  if (!(isValidPalindromeOutside("A man, a plan, a canal: Panama") === true)) throw new Error("Test Result Error");
  console.log("All test cases passed!");
}
catch(err) {
  console.log(err.stack)
}

File: leetcode/strings/valid_palindrome_two_pointers_from_outside.rb
# https://leetcode.com/problems/valid-palindrome/
# Given a string s, return true if it is a palindrome, or false otherwise.
# A phrase is a palindrome if, after converting all uppercase letters into
# lowercase letters and removing all non-alphanumeric characters, it reads
# the same forward and backward.

# Two pointers from outside:
# @param {String} s
# @return {Boolean}
def is_palindrome(s)
  s = s.gsub(/[^A-Za-z0-9]/, "").downcase
  
  return true if s.length <= 1
  
  left = 0
  right = s.length - 1
  
  while (left < right)
    if s[left] != s[right]
      return false
    end
    
    left += 1
    right -= 1
  end
  
  return true
end

if __FILE__ == $0
  begin
    raise RuntimeError unless is_palindrome("aabaa") == true
    raise RuntimeError unless is_palindrome("aabbaa") == true
    raise RuntimeError unless is_palindrome("abc") == false
    raise RuntimeError unless is_palindrome("a") == true 
    raise RuntimeError unless is_palindrome("") == true
    raise RuntimeError unless is_palindrome(" ") == true
    raise RuntimeError unless is_palindrome("A man, a plan, a canal: Panama") == true
    raise RuntimeError unless is_palindrome("race a car") == false
    raise RuntimeError unless is_palindrome("race car") == true 
  rescue RuntimeError => e
    puts e.message + ": " + e.backtrace.inspect
  else
    puts "All test cases passed!"
  end    
end